File: leetcode/medium/DeleteLeavesWithAGivenValue.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def removeLeafNodes(self, root: Optional[TreeNode], target: int) -> Optional[TreeNode]:
        """
        Removes all leaf nodes from a binary tree that have a specific target value.
        This process might cause parent nodes to become leaves, which are then also
        removed if they have the target value. This continues until no more such
        leaf nodes can be removed.

        The function uses a post-order traversal (DFS) approach. By processing
        the children first, we ensure that when we evaluate a node, its subtrees
        have already been pruned. If a node becomes a leaf after its children
        are processed and its value matches the target, it can be safely removed.

        Args:
            root: The root of the binary tree.
            target: The value of the leaf nodes to be removed.

        Returns:
            The root of the modified binary tree.
        """
        # Base case: if the node is None, there's nothing to do.
        if not root:
            return None

        # Recursively process the left and right subtrees.
        # The returned (potentially modified) subtrees are re-assigned.
        root.left = self.removeLeafNodes(root.left, target)
        root.right = self.removeLeafNodes(root.right, target)

        # After processing the children, check if the current node has become a leaf
        # and if its value matches the target.
        if not root.left and not root.right and root.val == target:
            # If so, this node should be removed. We do this by returning None
            # to its parent, which will effectively disconnect it from the tree.
            return None

        # If the node is not a target leaf, we keep it.
        return root

File: leetcode/medium/DeleteLeavesWithAGivenValue.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    /**
     * Removes all leaf nodes from the given binary tree that have a value equal to the target.
     * The process is recursive and works bottom-up (post-order traversal), meaning a non-leaf
     * node can become a new leaf node and subsequently be removed if its value matches the target
     * and its children were removed. 
     *
     * @param root The root of the binary tree.
     * @param target The value of leaf nodes to be removed.
     * @return The root of the resulting binary tree after the removal, or null if the entire tree is removed.
     * * @time_complexity O(N), where N is the number of nodes in the tree. Every node is visited exactly once.
     * @space_complexity O(H), where H is the height of the tree. This is due to the recursion stack. 
     * In the worst case (a skewed tree), H can be N. In the best/average case (a balanced tree), H is log(N).
     */
    public TreeNode removeLeafNodes(TreeNode root, int target) {
        // Base case: if the current node is null, return null.
        if (root == null) {
            return null;
        }

        // 1. Recurse on the left and right subtrees (Post-order traversal)
        // This ensures the children are processed and pruned *before* the current node.
        root.left = removeLeafNodes(root.left, target);
        root.right = removeLeafNodes(root.right, target);

        // 2. Check if the current node is now a leaf node with the target value
        // The children pointers will be null if they were removed in the recursive calls above.
        // A node is a leaf if both its left and right children are null.
        if (root.left == null && root.right == null && root.val == target) {
            // If it meets the criteria, remove it by returning null to its parent.
            return null;
        }

        // 3. If the node is not removed, return the node itself.
        return root;
    }
}